src/Disks/HDFS/DiskHDFS.cpp

```diff
@@ -0,0 +1,235 @@
+#include <Disks/HDFS/DiskHDFS.h>
+
+#include <Storages/HDFS/ReadBufferFromHDFS.h>
+#include <Storages/HDFS/WriteBufferFromHDFS.h>
+#include "ReadIndirectBufferFromHDFS.h"
+#include "WriteIndirectBufferFromHDFS.h"
+#include <IO/SeekAvoidingReadBuffer.h>
+#include <Common/checkStackSize.h>
+#include <Common/quoteString.h>
+#include <common/logger_useful.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+    extern const int LOGICAL_ERROR;
+    extern const int PATH_ACCESS_DENIED;
+    extern const int CANNOT_DELETE_DIRECTORY;
+    extern const int UNKNOWN_FORMAT;
+}
+
+
+DiskHDFS::DiskHDFS(
+    const String & disk_name_,
+    const String & hdfs_root_path_,
+    const String & metadata_path_,
+    const Poco::Util::AbstractConfiguration & config_,
+    size_t min_bytes_for_seek_)
+    : IDiskRemote(disk_name_, hdfs_root_path_, metadata_path_, "DiskHDFS")
+    , config(config_)
+    , min_bytes_for_seek(min_bytes_for_seek_)
+    , hdfs_builder(createHDFSBuilder(hdfs_root_path_, config))
+    , hdfs_fs(createHDFSFS(hdfs_builder.get()))
+{
+}
+
+
+std::unique_ptr<ReadBufferFromFileBase> DiskHDFS::readFile(const String & path, size_t buf_size, size_t, size_t, size_t, MMappedFileCache *) const
+{
+    auto metadata = readMeta(path);
+
+    LOG_DEBUG(log,
+        "Read from file by path: {}. Existing HDFS objects: {}",
+        backQuote(metadata_path + path), metadata.remote_fs_objects.size());
+
+    auto reader = std::make_unique<ReadIndirectBufferFromHDFS>(config, remote_fs_root_path, metadata, buf_size);
+    return std::make_unique<SeekAvoidingReadBuffer>(std::move(reader), min_bytes_for_seek);
+}
+
+
+std::unique_ptr<WriteBufferFromFileBase> DiskHDFS::writeFile(const String & path, size_t buf_size, WriteMode mode)
+{
+    bool exist = exists(path);
```

Comment:
`readOrCreateMetaForWriting` method may be used here.
Additionnaly this method can be modified to return flags mask (for HDFS).